src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3367,15 +3367,31 @@ void G1CollectedHeap::fill_with_dummy_object(HeapWord* start, HeapWord* end, boo
   region->fill_with_dummy_object(start, pointer_delta(end, start), zap);
 }
 
-void G1CollectedHeap::start_codecache_marking_cycle_if_inactive() {
+void G1CollectedHeap::start_codecache_marking_cycle_if_inactive(bool full_gc) {
+  // We can reach here with an active code cache marking cycle either because
+  // the previous G1 concurrent marking cycle
+  //
+  // (1) was aborted and a full gc is about to begin or
+  // (2) it was undone because the heap occupancy was below the threshold after
+  //     the initiating young gc (see G1ConcurrentMark::post_concurrent_undo_start())
+  //
+  // Especially in case (2) it is important that the active codecache cycle was not
+  // finished when executing the G1 undo operation. If it was (repeatedly) finished
+  // then nmethods with frames in continuation StackChunks could appear as not
+  // on stack because they were not marked as on stack in the undone concurrent marking.
+  //
+  // Also for (2) it is important to arm nmethod entry barriers even if no new
+  // code cache cycle is started. They are needed for a complete SATB. A full gc
+  // on the other hand doesn't need the barriers at all.
   if (!CodeCache::is_gc_marking_cycle_active()) {
-    // This is the normal case when we do not call collect when a
-    // concurrent mark is ongoing. We then start a new code marking
-    // cycle. If, on the other hand, a concurrent mark is ongoing, we
-    // will be conservative and use the last code marking cycle. Code
-    // caches marked between the two concurrent marks will live a bit
-    // longer than needed.
     CodeCache::on_gc_marking_cycle_start();
+  }
+  if (!full_gc) {
```

Comment:
Should this be else if?

Reply:
No, `!full_gc` means we reach here because concurrent marking is about to start and we want to arm unconditionally then. Never arm for the start of a full gc on the other hand.

Would it be clearer if I renamed `full_gc` to `concurrent_marking_start`?

Reply:
Yes, I think so. Otherwise, looks good. Thanks again!

Reply:
please use `concurrent_mark_start` (without the "ing"), or `in_concurrent_start_mark[_pause]`.